### ISSUE TITLE
COM interface methods are nicer to use if they return the last `out` parameter instead of void

### ISSUE THREAD

**Issue by jnm2** [microsoft/CsWin32#341]
It's nicer to be able to dot off returned values than to come up with names for `out var` parameters for void method calls.

It enables code that looks like this:

```cs
FileName = dialog.GetResult().GetDisplayName(SIGDN.SIGDN_DESKTOPABSOLUTEPARSING);
fileTypeIndex = (int)dialog.GetFileTypeIndex();
```

Instead of:

```cs
dialog.GetResult(out var shellItem);
shellItem.GetDisplayName(SIGDN.SIGDN_DESKTOPABSOLUTEPARSING, out var displayName);
FileName = new string(displayName);
// TODO: free displayName (my declaration returned `string` from GetDisplayName instead of PWSTR)
dialog.GetFileTypeIndex(out var uintFileTypeIndex);
fileTypeIndex = (int)uintFileTypeIndex;
```

My own declaration:

```cs
/// <summary>
/// https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nn-shobjidl_core-ifileopendialog
/// </summary>
[ComImport, Guid("D57C7288-D4AD-4768-BE02-9D969532D960")]
[InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
internal interface IFileOpenDialog
{
    // ...
    uint GetFileTypeIndex();
    // ...
    IShellItem GetResult();
    // ...
}
```

**Comment by AArnott** [comment]
We already do approximately what you're asking for (#120).
But we're not going to automatically move whatever the last out parameter is to be a return type. It's based on the metadata's use of `[RetVal]` on that last parameter. Those tend to come from the IDL or header files. If you search the metadata for uses of RetValAttribute you'll see a great many hits. So if your feedback is targeting specific APIs that lack it that you feel should have it, please file an issue for those APIs against the win32metadata repo.

**Comment by jnm2** [comment]
This is going to be a lot of API requests, since I'm not going to come across a void-returning method with a final `out` parameter that I'm going to appreciate.

**Comment by AArnott** [comment]
TBH I'm not sure we should deliver on those requests in the metadata repo either. Our goal is to produce the same methods that Windows defines, with friendly overloads that handle some type conversions and memory management in a .NET-friendly way. The methods should still closely resemble the native ones so that code conversion from C++ to C# is straightforward and documentation still applies.
Our goal is _not_ to make the APIs more like a BCL. Revising a method signature _just_ to make it more convenient (rather than for memory management or type compatibility reasons) seems to at least approach the scope line.

But you're welcome to file an issue on win32metadata of course and more folks can participate in the discussion there.

**Comment by jnm2** [comment]
The reason I bring it up is not to make APIs more like a BCL, but to provide parity with handwritten interop definitions.

**Comment by AArnott** [comment]
Another concern is I've learned as part of COM interop is that very subtle changes can surprisingly impact the ABI, which on some CPU architectures can break the runtime. I wonder whether an out parameter vs. a return value might break things. That's something else we'll need to verify before we generate something different.

**Comment by AArnott** [comment]
@weltkante @AaronRobinsonMSFT Do you have any concerns with rewriting COM interface methods with a signature like this:
```cs
HRESULT Some(out int foo);
```

which without preservesig would turn into this in C#:

```cs
void Some(out int foo);
```

into something even more refined like this:

```cs
int Some();
```

?

Any ABI concerns to be aware of?

As mentioned above, I know the IDL allows for `[retval]` annotations, which CsWin32 already honors to do the above transformation. But the suggestion in this issue is to perform this transformation even without this annotation.

**Comment by weltkante** [comment]
- you get exceptions instead of HRESULTs which can be a debugging/performance issue, especially when failures are common in a given API usage (I occasionally encounter that an API asks you to "try" something and will return a failure HRESULT if its not available)
- you cannot distinguish success results (S_OK vs. S_FALSE etc.)
- some out parameters are optional, but the `out` keyword is already the wrong projection for those anyways, so moving an out parameter into return parameter probably doesn't break those cases more than they are already broken now

In summary, the common case will probably profit from this transformation, but doing it in general breaks a minority of APIs and/or makes them much more painful to use.

**Comment by AaronRobinsonMSFT** [comment]
@weltkante's list is great. One thing that might impact the ABI in profound ways is for `out` value types that are moved to `retval`. 

From a COM purest standpoint, if the metadata doesn't have the `retval` it shouldn't be converted to be one. There are lots that can likely be done to manipulate the API to conform to C# patterns, but this one violates the API writer's intent. That may not be a good enough reason not to do it, but I could imagine niche cases, especially for value types, where this is not a good idea.

**Comment by jnm2** [comment]
I noticed that everything on @weltkante's list applies just as much to `void Some(out int foo);` as to `int Some();`, given that the preservesig definition would be `HRESULT Some(out int foo);`, so they would not be reasons to prefer `void Some(out int foo);` over `int Some();`.

@AaronRobinsonMSFT Do you think we should know more about the niche cases to consider whether (or how much) they affect CsWin32?

**Comment by AaronRobinsonMSFT** [comment]
> so they would not be reasons to prefer void Some(out int foo); over int Some();.

Agree for `int` and perhaps other primitive types. However, the distinction does break with the API author's definition _and_ could impact ABI for value types. I put a lot of value on how the API is defined, if it doesn't have `retval` it is changing the intent.

>Do you think we should know more about the niche cases to consider whether (or how much) they affect CsWin32?

Unfortunately, yes. I would be less concerned if CsWin32 could use .NET 6 APIs and wasn't limited to targeting .NET Standard. The reason for this is in .NET 6 we've added many interop niceties that help interop library developers tell the runtime additional context that makes the dispatch better (for example, [`CallConvMemberFunction`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.callconvmemberfunction)).

**Comment by AArnott** [comment]
Yes, the inefficiencies concerns are all based on `[PreserveSig]` which is a separate (and decided) question. That question was answered with "yes, preservesig by default but allow consumers to turn it off for an interface or individual method. Also include a list of APIs known to return `S_FALSE` and turn PreserveSig off by default for those."

And yes, I agree treating the API as if `[retval]` were present when it isn't violates the author's intent (assuming it was thought through).

But the key question I am looking for @AaronRobinsonMSFT and @weltkante to answer is around ABI compatibility. In what cases can we expect ABI incompatibilities when moving the last `out` parameter to be a return type on an otherwise void-returning method, while PreserveSig is _not_ present? @AaronRobinsonMSFT mentioned value types might break. Which value types? What about pointers to arbitrary memory or pointers to COM objects? 

If we're not sure the set, but are pretty sure it could present problems, I would proceed with the Won't Fix resolution of this suggestion. If we could get a defined set of problems and could avoid them, then we can consider this issue vs. staying true to the author's intent.

**Comment by AArnott** [comment]
Wow. As I am finding typical, this space is always more complex than I anticipate. Thank you, @AaronRobinsonMSFT. We may revisit this, but I think our first priority should be to get accurate APIs with ABI compatibility, and then to gradually get more user-friendly ones. As we still have existing ABI-compat bugs I think we should work on those before possibly introducing new bugs by reshaping these APIs.

Having this issue's discussion will be critical to getting it right eventually, if we do end up wanting to 'fix' this.
